compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -1990,31 +1983,39 @@ class Objects(using Context @constructorOnly):
    * Object access elision happens when the object access is used as a prefix
    * in `new o.C` and `C` does not need an outer.
    */
-  def resolveThis(target: ClassSymbol, thisV: Value, klass: ClassSymbol, elideObjectAccess: Boolean = false): Contextual[Value] = log("resolveThis target = " + target.show + ", this = " + thisV.show, printer, (_: Value).show) {
-    if target == klass then
-      thisV
-    else if target.is(Flags.Package) then
-      Package(target) // TODO: What is the semantics for package.this?
+  def resolveThis(target: ClassSymbol, thisV: Value, klass: ClassSymbol, elideObjectAccess: Boolean = false): Contextual[ThisValue] = log("resolveThis target = " + target.show + ", this = " + thisV.show, printer, (_: Value).show) {
+    def recur(scopeSet: ScopeSet): ThisValue =
+      if scopeSet == Env.NoEnv then
+        Bottom
+      else
+        val head = scopeSet.scopes.head
+        if head.isInstanceOf[Ref] then
+          val klass = head.asInstanceOf[Ref].klass
+          assert(scopeSet.scopes.forall(_.asInstanceOf[Ref].klass == klass), "Multiple possible outer class?")
+          if klass == target then
+            ValueSet(scopeSet.scopes.map(_.asInstanceOf[Ref]))
+          else
+            recur(scopeSet.scopes.map(_.outer).join)
+        else
+          recur(scopeSet.scopes.map(_.outer).join)
+    end recur
```

Comment:
For future: if possible, try to hide abstract domain details from the semantic logic. That would make the code more robust against refactoring and easier to understand.